README.md
Forecaster
==========

[![Gem](https://img.shields.io/gem/v/forecaster.svg)](https://rubygems.org/gems/forecaster)
[![Build Status](https://api.travis-ci.org/vinc/forecaster.svg?branch=master)](http://travis-ci.org/vinc/forecaster)
[![Code Climate](https://codeclimate.com/github/vinc/forecaster.svg)](https://codeclimate.com/github/vinc/forecaster)
[![Code Coverage](https://codecov.io/gh/vinc/forecaster/branch/master/graph/badge.svg)](https://codecov.io/gh/vinc/forecaster)

[Forecaster](https://github.com/vinc/forecaster) is a gem wrapping `wgrib2` to
fetch and read weather data directly from the Global Forecast System.

It comes with a library and a CLI allowing you to type commands like `forecast
for this afternoon in paris` in your terminal to get the latest weather
forecast.

[![asciicast](https://asciinema.org/a/193637.png)](https://asciinema.org/a/193637)


Installation
------------

```bash
gem install forecaster
```

Alternatively you can build the gem from its repository:

```bash
git clone git://github.com/vinc/forecaster.git
cd forecaster
gem build forecaster.gemspec
gem install forecaster-1.0.0.gem
```

In both cases you need to make sure that you have `wgrib2` present in your
system.

To install the `wgrib2` from source:

```bash
wget http://www.ftp.cpc.ncep.noaa.gov/wd51we/wgrib2/wgrib2.tgz
tar -xzvf wgrib2.tgz
cd grib2
export CC=gcc
export FC=gfortran
make
sudo cp wgrib2/wgrib2 /usr/local/bin/
```

Usage
-----

```ruby
require "forecaster"
```

To configure the gem:

```ruby
Forecaster.configure do |config|
  config.wgrib2_path = "/usr/local/bin/wgrib2"
  config.cache_dir = "/tmp/forecaster"
  config.records = {
    :temperature => ":TMP:2 m above ground:",
    :humidity    => ":RH:2 m above ground:",
    :pressure    => ":PRES:surface:"
  }
end
```

Forecaster saves large files containing the data of GFS runs from the NOAA
servers in the cache directory, but only the parts of the files containing
the records defined in the configuration will be downloaded.

You can find the list of available records [online][1] or by reading any
`.idx` files distributed along with the GFS files.

A record is identified by a variable and a layer separated by colon
characters. In the case of the temperature for example, those attributes
are `TMP` and `2 m above ground`. See the [documentation of wgrib2][2] for
more information.

To fetch a forecast:

```ruby
t = Time.now.utc # All the dates should be expressed in UTC
y = t.year       # year of GFS run
m = t.month      # month of GFS run
d = t.day        # day of GFS run
c = 0            # hour of GFS run (must be a multiple of 6)
h = 12           # hour of forecast (must be a multiple of 3)
forecast = Forecaster.fetch(y, m, d, c, h) # Forecaster::Forecast
```

To read the [record][1] of a forecast:

```ruby
res = forecast.read(:temperature, longitude: 48.1147, latitude: -1.6794) # String in Kelvin
val = res.to_f - 273.15 # Float in degree Celsius
```

[1]: http://www.nco.ncep.noaa.gov/pmb/products/gfs/gfs_upgrade/gfs.t06z.pgrb2.0p25.f006.shtml
[2]: http://www.cpc.ncep.noaa.gov/products/wesley/wgrib2/


Command line
------------

Forecaster has a command line tool that try to be smart:

    $ forecast for tomorrow afternoon in auckland
    GFS Weather Forecast

      Date:        2016-05-13
      Time:          12:00:00
      Zone:             +1200
      Latitude:         -36.8 °
      Longitude:        174.8 °

      Pressure:        1013.8 hPa
      Temperature:       21.7 °C
      Nebulosity:         0.0 %
      Humidity:          65.1 %
      Precipitation:      0.0 mm
      Wind:               8.0 m/s (NE)

But you can use it in a more verbose way:

    $ TZ=America/Los_Angeles forecast --time "2016-05-12 09:00:00" \
                                      --latitude "37.7749295" \
                                      --longitude "-122.4194155" \
                                      --debug
    Requested time:  2016-05-12 09:00:00 -0700
    GFS Run time:    2016-05-11 23:00:00 -0700
    Forecast time:   2016-05-12 08:00:00 -0700

    Downloading: 'http://www.ftp.ncep.noaa.gov/data/nccf/com/gfs/prod/gfs.20160
    51200/gfs.t00z.pgrb2.0p25.f015'
    Reading index file...
    Length: 4992281 (4.76M)

    100% [===========================================>] 696 KB/s Time: 00:00:07

    GFS Weather Forecast

      Date:        2016-05-12
      Time:          08:00:00
      Zone:             -0700
      Latitude:          37.8 °
      Longitude:       -122.4 °

      Pressure:        1013.5 hPa
      Temperature:       13.4 °C
      Nebulosity:         0.0 %
      Humidity:          89.7 %
      Precipitation:      0.0 mm
      Wind:               1.0 m/s (NE)

To use automatically the timezone of a location you will need to create
a free [GeoNames account][3] and export your username in an environment
variable:

    export GEONAMES_USERNAME=<username>

And while you're doing that, you can also export your favorite location
to avoid typing it every time:

    export FORECAST_LATITUDE=<latitude>
    export FORECAST_LONGITUDE=<longitude>

[3]: http://www.geonames.org/login


License
-------

Copyright (c) 2015-2018 Vincent Ollivier. Released under MIT.


CHANGELOG.md
# Changelog

## 1.0.2 (2018-10-22)

- Rename `trollop` dependency to `optimist`
- Rename "Cloud Cover" to "Nebulosity"

## 1.0.1 (2018-07-26)

- Fix `format` error

## 1.0.0 (2018-03-31)

- Colorize and refactor output
- Update dependencies
- Replace `Forecast#fetch_index` by `Forecast#fetch_ranges`
- Fix bug when time is not in UTC
- Drop support of ruby version < 2.1
- Increase code coverage

## 0.1.0 (2016-06-11)

- Add forecast executable
- Cache geolocation requests
- Replace curl wrapper with excon gem
- Fix cache path

## 0.0.2 (2016-06-09)

- Raise error when field could not be read in file

## 0.0.1 (2015-05-04)

- Publish initial version


spec/spec_helper.rb
require "simplecov"
SimpleCov.start

if ENV["CI"]
  require "codecov"
  SimpleCov.formatter = SimpleCov::Formatter::Codecov
end

require "forecaster"

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end
end


spec/forecast_spec.rb
require "tmpdir"
require "fileutils"

RSpec.describe Forecaster::Forecast do
  before do
    Forecaster.configure do |config|
      config.cache_dir = Dir.mktmpdir
      config.records = {
        :tmp => ":TMP:2 m above ground:" # Temperature
      }
    end

    t = Time.now - 86400
    y = t.year
    m = t.month
    d = t.day
    c = 0 # hour of GFS run
    h = 3 # hour of forecast
    @forecast = Forecaster::Forecast.new(y, m, d, c, h)
  end

  after do
    FileUtils.remove_entry_secure(Forecaster.configuration.cache_dir)
  end

  it "requires curl" do
    curl_path = Forecaster.configuration.curl_path
    out = `#{curl_path} --version`

    expect(out).to start_with("curl")
  end

  it "requires wgrib2" do
    wgrib2_path = Forecaster.configuration.wgrib2_path
    out = `#{wgrib2_path} -version`

    expect(out).to start_with("v0.2")
  end

  it "fetches a forecast" do
    expect(@forecast.fetched?).to be false
    @forecast.fetch
    expect(@forecast.fetched?).to be true
  end

  it "reads a forecast" do
    @forecast.fetch
    value = @forecast.read(:tmp, longitude: 48.1147, latitude: -1.6794)
    expect(value).to be_a(String)
    expect(value.to_i).to be_between(180, 340).inclusive # in Kelvin
  end
end


spec/forecaster_spec.rb
require "tmpdir"
require "fileutils"
require "timecop"

RSpec.describe Forecaster do
  before do
    Forecaster.configure do |config|
      config.cache_dir = Dir.mktmpdir
      config.records = {
        :tmp => ":TMP:2 m above ground:" # Temperature
      }
    end

    t = Time.now - 86_400
    @y = t.year
    @m = t.month
    @d = t.day
    @c = 0 # hour of GFS run
    @h = 3 # hour of forecast
  end

  after do
    FileUtils.remove_entry_secure(Forecaster.configuration.cache_dir)
  end

  it "requires wgrib2" do
    wgrib2_path = Forecaster.configuration.wgrib2_path
    out = `#{wgrib2_path} -version`

    expect(out).to start_with("v0.2")
  end

  it "fetches a forecast" do
    forecast = Forecaster.fetch(@y, @m, @d, @c, @h)

    expect(forecast.fetched?).to be true
  end

  describe Forecaster::Forecast do
    it "gets the time of the last GFS run" do
      # There are 4 GFS runs per day at 0h, 6h, 12h and 18h UTC.
      # They are available online 6 hours after the run.
      # See: http://www.nco.ncep.noaa.gov/pmb/products/gfs/

      t = Time.utc(2015, 1, 1, 0, 0)
      Timecop.freeze(t + 7 * 3600) do
        expect(Forecaster::Forecast.last_run_at).to eq(t)
      end
      Timecop.freeze(t + 13 * 3600) do
        expect(Forecaster::Forecast.last_run_at).to eq(t + 6 * 3600)
      end
    end

    it "create a forecast" do
      t = Time.utc(2015, 1, 1, 0, 0)
      Timecop.freeze(t + 7 * 3600) do
        # There is a forecast every 3 hours after a run for 384 hours.
        # See: http://www.nco.ncep.noaa.gov/pmb/products/gfs/

        #
        # Forecasts from an archived run
        #

        # exactly at the time of a forecast
        forecast = Forecaster::Forecast.at(t)
        expect(forecast.run_time).to eq(t - 6 * 3600)
        expect(forecast.time).to eq(t)

        # 1 hour after a forecast time
        forecast = Forecaster::Forecast.at(t + 1 * 3600)
        expect(forecast.run_time).to eq(t - 6 * 3600)
        expect(forecast.time).to eq(t)

        #
        # Forecasts from the last run
        #

        # exactly at the time of a forecast
        forecast = Forecaster::Forecast.at(t + 6 * 3600)
        expect(forecast.run_time).to eq(t)
        expect(forecast.time).to eq(t + 6 * 3600)

        # 1 hour after a forecast time
        forecast = Forecaster::Forecast.at(t + 4 * 3600)
        expect(forecast.run_time).to eq(t)
        expect(forecast.time).to eq(t + 3 * 3600)

        # 1 hour after a forecast time
        forecast = Forecaster::Forecast.at(t + 10 * 3600)
        expect(forecast.run_time).to eq(t)
        expect(forecast.time).to eq(t + 9 * 3600)

        # 1 hour after a forecast time
        forecast = Forecaster::Forecast.at(t + 49 * 3600)
        expect(forecast.run_time).to eq(t)
        expect(forecast.time).to eq(t + 48 * 3600)

        # TODO: Test with a time too far in the future
      end
    end

    it "fetches a forecast" do
      forecast = Forecaster::Forecast.new(@y, @m, @d, @c, @h)

      expect(forecast.fetched?).to be false
      forecast.fetch
      expect(forecast.fetched?).to be true
    end

    it "reads a forecast" do
      forecast = Forecaster::Forecast.new(@y, @m, @d, @c, @h)

      forecast.fetch
      value = forecast.read(:tmp, :longitude => 48.1147, :latitude => -1.6794)
      expect(value).to be_a(String)
      expect(value.to_i).to be_between(180, 340).inclusive # in Kelvin
    end

    it "fetches byte ranges from an index file" do
      forecast = Forecaster::Forecast.new(@y, @m, @d, @c, @h)

      ranges = forecast.fetch_ranges
      expect(ranges).to have_key(":TMP:2 m above ground:")

      range = ranges[":TMP:2 m above ground:"]
      expect(range.size).to be_between(1, 2)
      first_byte, last_byte = range
      expect(first_byte).to be_an(Integer)
      case range.size
      when 1
        expect(last_byte).to be_nil
      when 2
        expect(last_byte).to be_an(Integer)
      end
    end
  end

  describe Forecaster::CLI do
    describe "#start" do
      it "starts the CLI" do
        args = %w[for this afternoon --latitude 48.75 --longitude 02.25]
        env = { "TZ" => "Europe/Paris" }
        expect { Forecaster::CLI.start(args, env) }.to output(/Latitude.*48\.75/).to_stdout
      end
    end
  end
end


lib/forecaster/forecast.rb
require "fileutils"
require "excon"

# Fetch and read data from the Global Forecast System.
module Forecaster
  # Fetch and read a specific forecast from a GFS run.
  #
  # See: http://www.nco.ncep.noaa.gov/pmb/products/gfs/
  class Forecast
    def self.last_run_at
      # There is a new GFS run every 6 hours starting at midnight UTC, and it
      # takes approximately 3 to 5 hours before a run is available online, so
      # to be on the safe side we return the previous one.
      now = Time.now.utc
      run = Time.utc(now.year, now.month, now.day, (now.hour / 6) * 6)

      run - 6 * 3600
    end

    def self.at(time)
      # There is a forecast every 3 hours after a run for 384 hours.
      t = time.utc
      fct = Time.utc(t.year, t.month, t.day, (t.hour / 3) * 3)
      run = Time.utc(t.year, t.month, t.day, (t.hour / 6) * 6)
      run -= 6 * 3600 if run == fct

      last_run = Forecast.last_run_at
      run = last_run if run > last_run

      fct_hour = (fct - run) / 3600

      raise "Time too far in the future" if fct_hour > 384

      Forecast.new(run.year, run.month, run.day, run.hour, fct_hour)
    end

    def initialize(year, month, day, run_hour, fct_hour)
      @year = year
      @month = month
      @day = day
      @run_hour = run_hour
      @fct_hour = fct_hour
    end

    def run_time
      Time.utc(@year, @month, @day, @run_hour)
    end

    def time
      run_time + @fct_hour * 3600
    end

    def dirname
      subdir = format("%04d%02d%02d%02d", @year, @month, @day, @run_hour)
      File.join(Forecaster.configuration.cache_dir, subdir)
    end

    def filename
      format("gfs.t%02dz.pgrb2.0p25.f%03d", @run_hour, @fct_hour)
    end

    def url
      server = Forecaster.configuration.server
      subdir = format("gfs.%04d%02d%02d%02d", @year, @month, @day, @run_hour)
      format("%s/%s/%s", server, subdir, filename)
    end

    def fetched?
      File.exist?(File.join(dirname, filename))
    end

    # This method will save the forecast file in the cache directory.
    # But only the parts of the file containing the fields defined in
    # the configuration will be downloaded.
    def fetch
      return if fetched?

      ranges = fetch_ranges

      # Select which byte ranges to download
      records = Forecaster.configuration.records.values
      filtered_ranges = records.map { |k| ranges[k] }

      fetch_grib2(filtered_ranges)
    end

    # Fetch the index file of a GRIB2 file containing the data of a forecast.
    #
    # Returns a hashmap of every records in the file with their byte ranges.
    #
    # With this method we can avoid downloading unnecessary parts of the GRIB2
    # file by matching the records defined in the configuration. It can also
    # be used to set up the later.
    def fetch_ranges
      begin
        res = Excon.get("#{url}.idx")
      rescue Excon::Errors::Error
        raise "Download of '#{url}.idx' failed"
      end
      lines = res.body.lines.map { |line| line.split(":") }
      lines.each_index.each_with_object({}) do |i, ranges|
        # A typical line (before the split on `:`) looks like this:
        # `12:4593854:d=2016051118:TMP:2 mb:9 hour fcst:`
        line = lines[i]
        next_line = lines[i + 1] # NOTE: Will be `nil` on the last line

        # The fourth and fifth fields constitue the key to identify the records
        # defined in `Forecaster::Configuration`.
        key = ":#{line[3]}:#{line[4]}:"

        # The second field is the first byte of the record in the GRIB2 file.
        ranges[key] = [line[1].to_i]

        # To get the last byte we need to read the next line.
        # If we are on the last line we won't be able to get the last byte,
        # but we don't need it according to the section 14.35.1 Byte Ranges
        # of RFC 2616.
        ranges[key] << next_line[1].to_i - 1 if next_line
      end
    end

    def fetch_grib2(ranges, progress_block: nil)
      FileUtils.mkpath(dirname)
      path = File.join(dirname, filename)

      streamer = lambda do |chunk, remaining, total|
        File.open(path, "ab") { |f| f.write(chunk) }
        progress_block.call(total - remaining, total) if progress_block
      end

      byte_ranges = ranges.map { |r| r.join("-") }.join(",")
      headers = { "Range" => "bytes=#{byte_ranges}" }

      begin
        Excon.get(url, :headers => headers, :response_block => streamer)
      rescue Excon::Errors::Error => e
        File.delete(path)
        raise "Download of '#{url}' failed: #{e}"
      end
    end

    def read(field, latitude: 0.0, longitude: 0.0)
      wgrib2 = Forecaster.configuration.wgrib2_path
      record = Forecaster.configuration.records[field]
      path = File.join(dirname, filename)

      raise "'#{path}' not found" unless File.exist?(path)

      coords = "#{longitude} #{latitude}"
      output = `#{wgrib2} #{path} -lon #{coords} -match "#{record}"`

      raise "Could not read '#{record}' in '#{path}'" if output.empty?

      fields = output.split("\n").first.split(":")
      params = Hash[*fields.last.split(",").map { |s| s.split("=") }.flatten]

      params["val"]
    end
  end
end


lib/forecaster/configuration.rb
# Fetch and read data from the Global Forecast System.
module Forecaster
  # Configure how to fetch and read from a forecast file.
  class Configuration
    attr_accessor :server, :cache_dir, :curl_path, :wgrib2_path, :records

    def initialize
      @server = "http://www.ftp.ncep.noaa.gov/data/nccf/com/gfs/prod"
      @cache_dir = "/tmp/forecaster"
      @wgrib2_path = "wgrib2"

      # See: http://www.nco.ncep.noaa.gov/pmb/products/gfs/gfs_upgrade/gfs.t06z.pgrb2.0p25.f006.shtml
      # See: http://www.cpc.ncep.noaa.gov/products/wesley/wgrib2/
      # Use `variable` and `level` attributes separated by colons to identify
      # the records to download and read.
      @records = {
        :prate => ":PRATE:surface:",
        :tmp   => ":TMP:2 m above ground:",
        :ugrd  => ":UGRD:10 m above ground:",
        :vgrd  => ":VGRD:10 m above ground:",
        :tcdc  => ":TCDC:entire atmosphere:"
      }
    end

    def self.configure(options)
      options.each do |option, value|
        self[option] = value
      end
    end
  end
end


lib/forecaster/cli.rb
require "yaml/store"
require "optimist"
require "chronic"
require "timezone"
require "geocoder"
require "ruby-progressbar"
require "rainbow"

require "forecaster"

# Fetch and read data from the Global Forecast System.
module Forecaster
  # Command line interface printing the forecast for a time and a location.
  class CLI
    include Singleton # TODO: Find how best to organize CLI class

    def self.start(args, env)
      instance.start(args, env)
    end

    def initialize
      @store = nil
    end

    def start(args, env)
      opts = parse(args)

      configure(opts)

      cache_file = File.join(Forecaster.configuration.cache_dir, "forecast.yml")
      @store = YAML::Store.new(cache_file)

      puts Rainbow("GFS Weather Forecast").bright
      puts

      lat, lon = get_location(opts, env)

      Optimist.die("Could not parse location") if lat.nil? || lon.nil?

      ENV["TZ"] = get_timezone(lat, lon, env) || env["TZ"]
      time = get_time(opts)
      forecast = get_forecast(time, opts)
      print_forecast(forecast, lat, lon)
      ENV["TZ"] = env["TZ"] # Restore TZ
    end

    # Parse command line options
    def parse(args)
      opts = Optimist.options(args) do
        usage           "for <time> in <location>"
        version         "Forecaster v#{Forecaster::VERSION}"
        opt :time,      "Set time in any format",  :type => String
        opt :location,  "Set location name",       :type => String
        opt :latitude,  "Set latitude in degree",  :type => Float, :short => "a"
        opt :longitude, "Set longitude in degree", :type => Float, :short => "o"
        opt :cache,     "Set cache directory",     :default => "/tmp/forecaster"
        opt :debug,     "Enable debug mode"
      end

      cmd_opts = { :time => [], :location => [] }
      key = :time
      args.each do |word|
        case word
        when "for"
          key = :time
        when "in"
          key = :location
        else
          cmd_opts[key] << word
        end
      end
      opts[:time] ||= cmd_opts[:time].join(" ")
      opts[:location] ||= cmd_opts[:location].join(" ")

      opts
    end

    # Configure gem
    def configure(opts)
      Forecaster.configure do |config|
        config.cache_dir = opts[:cache]
        config.records = {
          :prate => ":PRATE:surface:",
          :pres  => ":PRES:surface:",
          :rh    => ":RH:2 m above ground:",
          :tmp   => ":TMP:2 m above ground:",
          :ugrd  => ":UGRD:10 m above ground:",
          :vgrd  => ":VGRD:10 m above ground:",
          :tcdc  => ":TCDC:entire atmosphere:"
        }
      end
      FileUtils.mkpath(Forecaster.configuration.cache_dir)
    end

    # Get location
    def get_location(opts, env)
      if (opts[:location] || "").length > 0
        @store.transaction do
          if opts[:debug]
            putf_debug("Geolocalizing", opts[:location], "'%s'")
          end

          key = "geocoder:#{opts[:location]}"
          lat, lon = @store[key] ||= geolocalize(opts[:location])

          if opts[:debug]
            if lat && lon
              putf_debug("Location", lat, "%05.2f, %05.2f", optional: lon)
            else
              puts Rainbow("Location not found").red
            end
            puts
          end

          [lat, lon]
        end
      elsif opts[:latitude] && opts[:longitude]
        [opts[:latitude], opts[:longitude]]
      else
        [env["FORECAST_LATITUDE"], env["FORECAST_LONGITUDE"]]
      end
    end

    # Get timezone
    def get_timezone(lat, lon, env)
      tz = nil
      if env["GEONAMES_USERNAME"]
        Timezone::Lookup.config(:geonames) do |config|
          config.username = env["GEONAMES_USERNAME"]
        end
        @store.transaction do
          key = "timezone:#{lat}:#{lon}"
          tz = @store[key] || @store[key] = Timezone.lookup(lat, lon).name
        end
      end

      tz
    end

    # Get time
    def get_time(opts)
      if opts[:time]
        # TODO: Look for a timestamp first
        time = Chronic.parse(opts[:time])
        Optimist.die(:time, "could not be parsed") if time.nil?
        time.utc
      else
        Time.now.utc
      end
    end

    # Get forecast
    def get_forecast(time, opts)
      forecast = Forecast.at(time)

      if opts[:debug]
        putf_debug("Requested time", time.localtime,              "%s")
        putf_debug("GFS run time",   forecast.run_time.localtime, "%s")
        putf_debug("Forecast time",  forecast.time.localtime,     "%s")
        puts
      end

      unless forecast.fetched?
        if opts[:debug]
          putf_debug("Downloading", forecast.url, "'%s'")

          putf_debug("Reading index file",  "", "")
          records = Forecaster.configuration.records.values
          ranges = forecast.fetch_ranges
          ranges = records.map { |k| ranges[k] } # Filter ranges

          filesize = ranges.reduce(0) do |acc, (first, last)|
            acc + last - first # FIXME: `last == nil` on last range of index file
          end
          n = (filesize.to_f / (1 << 20)).round(2)
          putf_debug("Length", filesize, "%d (%.2fM)", optional: n)
          puts

          progressbar = ProgressBar.create(
            :format => "%p%% [%b>%i] %r KB/s %e",
            :rate_scale => lambda { |rate| rate / 1024 }
          )

          progress_block = lambda do |progress, total|
            progressbar.total = total
            progressbar.progress = progress
          end

          forecast.fetch_grib2(ranges, :progress_block => progress_block)

          progressbar.finish
          puts
        else
          forecast.fetch # That's a lot easier ^^
        end
      end

      forecast
    end

    # Print forecast
    def print_forecast(forecast, lat, lon)
      putf("Date", forecast.time.localtime.strftime("%Y-%m-%d"), "%s")
      putf("Time", forecast.time.localtime.strftime("%T"),       "%s")
      putf("Zone", forecast.time.localtime.strftime("%z"),       "%s")

      # Coordinates rounded to the precision of the GFS model
      putf("Latitude",  (lat / 0.25).round / 4.0, "%05.2f °")
      putf("Longitude", (lon / 0.25).round / 4.0, "%05.2f °")
      puts

      tmp   = forecast.read(:tmp,   :latitude => lat, :longitude => lon).to_f
      ugrd  = forecast.read(:ugrd,  :latitude => lat, :longitude => lon).to_f
      vgrd  = forecast.read(:vgrd,  :latitude => lat, :longitude => lon).to_f
      prate = forecast.read(:prate, :latitude => lat, :longitude => lon).to_f
      rh    = forecast.read(:rh,    :latitude => lat, :longitude => lon).to_f
      tcdc  = forecast.read(:tcdc,  :latitude => lat, :longitude => lon).to_f
      pres  = forecast.read(:pres,  :latitude => lat, :longitude => lon).to_f

      temperature    = tmp - 273.15
      wind_direction = (270 - Math.atan2(ugrd, vgrd) * 180 / Math::PI) % 360
      wind_speed     = Math.sqrt(ugrd**2 + vgrd**2)
      precipitation  = prate * 3600
      humidity       = rh
      nebulosity     = tcdc
      pressure       = pres / 100.0

      wdir = compass_rose(wind_direction)

      putf("Pressure",      pressure,      "%.0f hPa")
      putf("Temperature",   temperature,   "%.0f °C")
      putf("Nebulosity",    nebulosity,    "%.0f %%")
      putf("Humidity",      humidity,      "%.0f %%")
      putf("Precipitation", precipitation, "%.1f mm")
      putf("Wind",          wind_speed,    "%.1f m/s (%s)", optional: wdir)
    end

    def putf(name, value, fmt, optional: "", color: :cyan)
      left_column = Rainbow(format("  %-20s", name)).color(color)
      right_column = Rainbow(format(fmt, value, optional))
      puts "#{left_column} #{right_column}"
    end

    def putf_debug(name, value, fmt, optional: "")
      putf(name, value, fmt, optional: optional, color: :yellow)
    end

    def compass_rose(degree)
      case degree
      when   0...45  then "N"
      when  45...90  then "NE"
      when  90...135 then "E"
      when 135...180 then "SE"
      when 180...225 then "S"
      when 225...270 then "SW"
      when 270...315 then "W"
      else                "NW"
      end
    end

    def geolocalize(location)
      Geocoder.configure(:timeout => 10)
      res = Geocoder.search(location).first
      [res.latitude, res.longitude] if res
    end
  end
end
